[PyTorch][Edge] Add backport_toversion to export expected version model

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56855 [PyTorch][Edge] Add backport_toversion to export expected version model**
* #56802 [PyTorch Edge] Add backport to export old bytecode models
* #56002 [PyTorch Edge] Reuse constant table from ts in bytecode
* #56801 [PyTorch][Edge] Add api to get bytecode model version

Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

Commit:
[PyTorch][Edge] Add backport_toversion to export expected version model



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"






Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

iseeyuan commented:
As discussed offline, could we
1. Pull #56801 out of this stack and land it independently. Consider to add "from model file" or something like that to differentiate this PR from the top diff D27987811 (you may not have exported it yet).
2. Pull the top diff D27987811 to get version number from runtime, and land it independently. 1 and 2 are independent as well and it's not necessary to put them in the same diff. 
3. In the back port function, please reuse `readArchiveAndTensors` method in `serialization/import.h`, which is already a public function. It may avoid some copy-pastes. 
Thanks!

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

cccclai commented:
> As discussed offline, could we
> 
> 1. Pull #56801 out of this stack and land it independently. Consider to add "from model file" or something like that to differentiate this PR from the top diff D27987811 (you may not have exported it yet).
> 2. Pull the top diff D27987811 to get version number from runtime, and land it independently. 1 and 2 are independent as well and it's not necessary to put them in the same diff.
> 3. In the back port function, please reuse `readArchiveAndTensors` method in `serialization/import.h`, which is already a public function. It may avoid some copy-pastes.
>    Thanks!

1. Have #56801 at the bottom of the stack and rename `get_bytecode_version` to `get_model_bytecode_version`. The function `get_model_bytecode_version()` is needed by `backport` and `backport_to_version` apis. Just to clarify, the api for `get_model_bytecode_version` is not needed, but the function under the hood is needed. We can land this pr #56801 first. 
2. The api `get_runtime_bytecode_version()` is pulled as an independent pr #56948.
3. update #56802 with `readArchiveAndTensors` in the latest commit.

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add backport_toversion to export expected version model"


Add api `back_to_version(model_vn, version=i) -> model_vi`. Under the hood, add iteration to `backport(model_vn) -> model_vn-1`

## Test
```
python test/mobile/test_bytecode.py
```
Test with prod models (v5)



Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685/)

Differential Revision: [D27984685](https://our.internmc.facebook.com/intern/diff/D27984685)

[ghstack-poisoned]
